File: visualiser/tests/manual/neurokit_compare/check_edasymp.py
"""
EDASymp (0.045-0.25 Hz spectral sympathetic index) comparison - NeuroKit2
reference side.

Runs NeuroKit2's own `nk.eda_sympathetic(..., method='posada2016')` on each
track and emits the scalar indexes as JSON on stdout, keyed by track name.
The JS side (check_edasymp.js) computes BioMapping's SpectralEDA scalar on
the same raw signal and reports per-track ratio + cross-track correlation.

Why the comparison is a scalar (not a per-sample series): NeuroKit2's
posada2016 implementation returns ONE number per recording - the trapezoidal
integral of the Welch PSD (nperseg=128 @ 2 Hz = 64 s, 50 % overlap, periodic
Blackman window, nfft=256) over [0.045, 0.25) Hz. BioMapping's standalone
metric is the sliding-window extension of that same integral; the whole-
recording scalar is what must agree with NeuroKit2.

See docs/edasymp_spectral_investigation_proposal.md.

Usage: python3 check_edasymp.py track.csv [more.csv ...]
       (normally invoked via check_edasymp.sh, not directly)
"""
import json
import sys
from pathlib import Path

import pandas as pd
import numpy as np
import neurokit2 as nk

sys.path.insert(0, str(Path(__file__).parent))
from run_neurokit import to_microsiemens  # noqa: E402


def process(csv_path):
    df = pd.read_csv(csv_path, comment='#')
    eda = to_microsiemens(df['gsr_raw'].astype(float), 'gsr_raw')

    ts = df['timestamp'].values
    dt = pd.Series(ts).diff()
    dt = dt[dt > 0]
    sampling_rate = 1.0 / dt.mean() if len(dt) else 10.0

    res = nk.eda_sympathetic(eda, sampling_rate=sampling_rate, method='posada2016')

    symp = res['EDA_Sympathetic']
    symp_n = res['EDA_SympatheticN']
    return {
        'n_samples': len(df),
        'sampling_rate': float(sampling_rate),
        'eda_sympathetic': None if pd.isna(symp) else float(symp),
        'eda_sympathetic_n': None if pd.isna(symp_n) else float(symp_n),
    }


def main():
    if len(sys.argv) < 2:
        print('Usage: python3 check_edasymp.py track.csv [more.csv ...]', file=sys.stderr)
        sys.exit(1)

    out = {}
    for path in sys.argv[1:]:
        out[Path(path).stem] = process(path)
    json.dump(out, sys.stdout)


if __name__ == '__main__':
    main()

File: visualiser/tests/manual/neurokit_compare/check_edasymp.js
/**
 * EDASymp comparison (JS side) — see check_edasymp.py.
 *
 * Loads BioMapping's REAL SpectralEDA module (src/signal/spectral_eda.js),
 * re-reads each raw track CSV with the same µS unit conversion the Python
 * side applies (run_neurokit.to_microsiemens), and reports per-track
 * ratio + cross-track Pearson r against NeuroKit2's
 * nk.eda_sympathetic(posada2016) scalar.
 *
 * Usage (normally via check_edasymp.sh, not directly):
 *   node check_edasymp.js <python.json> <tracksdir>
 */
'use strict';

const fs = require('fs');
const path = require('path');

const { SpectralEDA } = require('../../../src/signal/spectral_eda.js');

// Mirrors run_neurokit.to_microsiemens (same unit detection as GSRCSVParser).
function toMicrosiemens(vals) {
  const avg = vals.reduce((s, v) => s + v, 0) / vals.length;
  if (avg > 50000) return vals.map(v => (v > 0 ? 1e6 / v : 0));
  if (avg > 100 && avg <= 50000) return vals.map(v => v / 1000);
  return vals;
}

function loadTrack(csvPath) {
  const lines = fs.readFileSync(csvPath, 'utf8').split('\n');
  const headerIdx = lines.findIndex(l => /^timestamp,/.test(l));
  if (headerIdx < 0) throw new Error(`No timestamp header in ${csvPath}`);
  const header = lines[headerIdx].split(',');
  const gsrIdx = header.indexOf('gsr_raw');
  const timeIdx = header.indexOf('timestamp');
  const times = [], vals = [];
  for (let i = headerIdx + 1; i < lines.length; i++) {
    const p = lines[i].split(',');
    if (p.length <= gsrIdx) continue;
    const t = parseFloat(p[timeIdx]);
    const v = parseFloat(p[gsrIdx]);
    if (!isNaN(t) && !isNaN(v)) { times.push(t); vals.push(v); }
  }
  return { us: toMicrosiemens(vals), times };
}

function samplingRateHz(times) {
  if (times.length < 2) return 10;
  const dt = times[times.length - 1] - times[0];
  return dt > 0 ? (times.length - 1) / dt : 10;
}

function pearson(a, b) {
  const n = Math.min(a.length, b.length);
  let sa = 0, sb = 0, sab = 0, sa2 = 0, sb2 = 0;
  for (let i = 0; i < n; i++) {
    sa += a[i]; sb += b[i]; sab += a[i] * b[i]; sa2 += a[i] * a[i]; sb2 += b[i] * b[i];
  }
  const cov = sab / n - (sa / n) * (sb / n);
  const va = sa2 / n - (sa / n) * (sa / n);
  const vb = sb2 / n - (sb / n) * (sb / n);
  return cov / Math.sqrt(va * vb);
}

const [, , jsonPath, tracksDir] = process.argv;
if (!jsonPath || !tracksDir) {
  console.error('Usage: node check_edasymp.js <python.json> <tracksdir>');
  process.exit(1);
}

const py = JSON.parse(fs.readFileSync(jsonPath, 'utf8'));
const names = Object.keys(py);

const ours = [], theirs = [];
console.log('=== EDASymp (0.045-0.25 Hz) agreement — ours (SpectralEDA) vs NeuroKit2 posada2016 ===');
for (const name of names) {
  const ref = py[name];
  if (ref.eda_sympathetic == null) {
    console.log(`  ${name}: NeuroKit2 returned NaN (signal <= 64 s) — skipped`);
    continue;
  }
  const track = loadTrack(path.join(tracksDir, name + '.csv'));
  const fs = samplingRateHz(track.times);
  const s = SpectralEDA.computeScalar(track.us, fs);
  const ratio = s.sympathetic / ref.eda_sympathetic;
  ours.push(s.sympathetic); theirs.push(ref.eda_sympathetic);
  console.log(`  ${name}: NK=${ref.eda_sympathetic.toFixed(6)} ours=${s.sympathetic.toFixed(6)} ratio=${ratio.toFixed(4)}  (n=${ref.n_samples}, fs=${ref.sampling_rate.toFixed(2)})`);
}

if (ours.length >= 2) {
  const r = pearson(ours, theirs);
  console.log(`\n  cross-track Pearson r = ${r.toFixed(6)} over ${ours.length} tracks`);
}
